File: Java/Eta/TestTools/QATools/wlconsumer-MutiThreaded-001/WatchlistConsumerConfig.java
package com.refinitiv.eta.valueadd.examples.watchlistconsumer;

import java.util.ArrayList;
import java.util.List;

import com.refinitiv.eta.codec.CodecFactory;
import com.refinitiv.eta.codec.CodecReturnCodes;
import com.refinitiv.eta.codec.MsgKey;
import com.refinitiv.eta.codec.MsgKeyFlags;
import com.refinitiv.eta.codec.State;
// APIQA
import com.refinitiv.eta.perftools.common.XmlItemInfoList;
import com.refinitiv.eta.shared.CommandLine;
import com.refinitiv.eta.rdm.DomainTypes;
import com.refinitiv.eta.transport.ConnectionTypes;
import com.refinitiv.eta.valueadd.examples.common.ConnectionArg;
import com.refinitiv.eta.valueadd.examples.common.ItemArg;

public class WatchlistConsumerConfig 
{
	private String publisherId;
	private String publisherAddress;
	private boolean enableEncrypted;
	private boolean enableHttp;
	private boolean cacheOption;
	private int cacheInterval;
	
	List<ConnectionArg> connectionList = new ArrayList<ConnectionArg>();
	   
    // default server host name
    private static final String defaultSrvrHostname = "localhost";
    
    // default server port number
    private static final String defaultSrvrPortNo = "14002";
    
    // default service name
    private static final String defaultServiceName = "DIRECT_FEED";
    
    // default item name
    private static final String defaultItemName = "TRI.N";
         	
    private static final int defaultRuntime = 600;    
	
	private int MAX_ITEMS = 50000;
	private int ITEMS_MIN_STREAM_ID = 5;	
	
	private ArrayList<ItemInfo> itemList = new ArrayList<ItemInfo>();
	private ArrayList<ItemInfo> providedItemList = new ArrayList<ItemInfo>();
	//APIQA
	private String itemFile;
	private int itemCount;
	
	class ItemInfo
	{
		int	domain;		
		String name;		
		int	streamId;		
		boolean	symbolListData;	
		boolean isPrivateStream;
		boolean isBatchStream;
        State state = CodecFactory.createState();
									
		public int domain()
		{
			return domain;
		}
		public void domain(int domainType)
		{
			this.domain = domainType;
		}
		public String name()
		{
			return name;
		}
		public void name(String name) 
		{
			this.name = name;
		}
		public int streamId()
		{
			return streamId;
		}
		public void streamId(int streamId)
		{
			this.streamId = streamId;
		}
		public boolean symbolListData()
		{
			return symbolListData;
		}
		public void symbolListData(boolean symbolListData)
		{
			this.symbolListData = symbolListData;
		}
		public boolean isPrivateStream() 
		{
			return isPrivateStream;
		}
		public void privateStream(boolean isPrivateStream)
		{
			this.isPrivateStream = isPrivateStream;
		}		
		public boolean isBatchStream() 
		{
			return isBatchStream;
		}
		public void batchStream(boolean isBatchStream)
		{
			this.isBatchStream = isBatchStream;
		}					
		public State state() 
		{
			return state;
		}
		public void state(State state)
		{
			this.state = state;
		}					
	}
	
	public boolean init(String[]args)
	{
		boolean ret;
	       	
		
		if ((ret = parseArgs(args)) == false )
		{
			return ret;	        	
		}
	//APIQA	
		List<ItemArg> itemList;
		if (itemFile == null)
		{
			List<ConnectionArg> connectionList = connectionList();
			itemList = connectionList.get(0).itemList();
		}
		else
			itemList = parseItemsFromFile();
		
		if (itemList == null)
		{
			System.out.println("Config error... ");
			System.exit(-1);
		}
		
		for ( ItemArg itemArg : itemList)
		{		
			addItem(itemArg.itemName(), itemArg.domain(), itemArg.symbolListData(), itemArg.enablePrivateStream());
		}
			
		initialAllConnectionArgs();
		
		return true;
	}
	
	
	public void addItem(String itemName, int domainType, boolean isSymbolList, boolean isPrivate)
	{
		ItemInfo itemInfo = new ItemInfo();
		itemInfo.domain(domainType);
		itemInfo.name(itemName);
		itemInfo.symbolListData(isSymbolList);
		itemInfo.privateStream(isPrivate);
		itemInfo.streamId(ITEMS_MIN_STREAM_ID + itemList.size());		
		itemList.add(itemInfo);
		
		if (itemList.size() >= MAX_ITEMS)
		{
			System.out.println("Config Error: Example only supports up to %d items " + MAX_ITEMS);
			System.exit(-1);
		}		
	}
		
	public ItemInfo getItemInfo(int streamId)
	{
		if (streamId > 0)
		{
			if (streamId >= ITEMS_MIN_STREAM_ID && streamId < itemList.size() + ITEMS_MIN_STREAM_ID)
				return itemList.get(streamId - ITEMS_MIN_STREAM_ID);
			else
				return null;
		}
		else if (streamId < 0)
		{
			for(int i = 0; i < providedItemList.size(); ++i)
			{
				if (providedItemList.get(i).streamId() == streamId)
					return providedItemList.get(i);
			}
			return null;
		}
		else
			return null;
	}

	public ItemInfo addProvidedItemInfo(int streamId, MsgKey msgKey, int domainType)
	{
		ItemInfo item = null;

		/* Check if item is already present. */
		for(int i = 0; i < providedItemList.size(); ++i)
		{
			if (providedItemList.get(i).streamId == streamId)
			{
				item = providedItemList.get(i);
				break;
			}
		}

		/* Add new item. */
		if (item == null)
		{
			if (providedItemList.size() == MAX_ITEMS)
			{
				System.out.println("Too many provided items.\n");
				return null;
			}
			item = new ItemInfo();
			providedItemList.add(item);
		}
	
		if ((msgKey.flags() & MsgKeyFlags.HAS_NAME)> 0)
		{
			item.name = msgKey.name().toString();
		}
		else
		{
			item.name = null;		
		}

		item.streamId(streamId);
		item.domain(domainType);

		return item;
	}

	public void removeProvidedItemInfo(ItemInfo item)
	{
		int i = 0;
		boolean found = false;
		for(i = 0; i < providedItemList.size(); ++i)
		{
			if (providedItemList.get(i).streamId == item.streamId)
			{
				found = true;
				break;
			}
		}
		if ( found ) 
		{
			providedItemList.remove(i);
		}
		return;					
	}

	public boolean parseArgs(String[] args)
	{        
		for (int i = 0; i < args.length; i++)
		{
			if (args[i].contains("runtime"))
			{
				args[i] = args[i].replace("runtime", "runTime");
			}
			if (args[i].contains("uname"))
			{
				args[i] = args[i].replace("uname", "u");
			}
	        if (args[i].contains("connectionType"))
	        {
	            args[i] = args[i].replace("connectionType", "c");
            }
//APIQA	        
	        if (args[i].contains("server"))
	        	parseSocketConnectionArgs(args, i) ;
	    	if (args[i].contains("itemFile"))
	    		itemFile = args[++i];
	    	if (args[i].contains("itemCount"))
	    		itemCount = Integer.parseInt(args[++i]);
		}
		
        addCommandLineArgs();
        try
        {
            CommandLine.parseArgs(args);
            
        	if (connectionList.size() == 0)
          		connectionList.add(new ConnectionArg());           
    		
        	ConnectionArg connectionArg = connectionList.get(0);
        	
    		String connectionType = CommandLine.value("c");
       	    if (connectionType.equals("socket"))
    	    {
    	    	connectionArg.connectionType(ConnectionTypes.SOCKET); 
    	    }
       	    else if (connectionType.equals("encrypted"))
    	    {
    	    	connectionArg.connectionType(ConnectionTypes.ENCRYPTED);
    			enableEncrypted = true;
    		}  
      	    else if (connectionType.equals("http"))
    	    {
    	    	connectionArg.connectionType(ConnectionTypes.HTTP);
    	    	enableHttp = true;
    	    }
			//APIQA
       	    if (connectionArg.service() == null)
       	    	connectionArg.service(serviceName());
        		
	       	 if (connectionArg.hostname() == null)
	    		connectionArg.hostname(CommandLine.value("h"));
	       	 if (connectionArg.port()  == null)
	    		connectionArg.port(CommandLine.value("p"));
            connectionArg.interfaceName(CommandLine.value("if"));

    		if (CommandLine.hasArg("tsServiceName"))
    		{
    			connectionArg.tsService(CommandLine.value("tsServiceName"));
    		}
    		            
    		if (CommandLine.hasArg("tunnel"))
    		{
    			connectionArg.tunnel(CommandLine.booleanValue("tunnel"));
    		}
    		
    		if (CommandLine.hasArg("tsAuth"))
    			connectionArg.tunnelAuth(CommandLine.booleanValue("tsAuth"));
    		if (CommandLine.hasArg("tsDomain"))
    			connectionArg.tunnelDomain(CommandLine.intValue("tsDomain"));

    		List<ItemArg> itemList = new ArrayList<ItemArg>();
    		
    		List<String> itemNames = CommandLine.values("mp");    		
    		if ( itemNames != null && itemNames.size() > 0 )
    			parseItems(itemNames, DomainTypes.MARKET_PRICE, false, false, itemList);
    		
       		itemNames = CommandLine.values("mbo");
       		if ( itemNames != null && itemNames.size() > 0 )
       			parseItems(itemNames, DomainTypes.MARKET_BY_ORDER, false, false, itemList);
    		
       		itemNames = CommandLine.values("mbp");
       		if ( itemNames != null && itemNames.size() > 0 )
       			parseItems(itemNames, DomainTypes.MARKET_BY_PRICE, false, false, itemList);
    		
       		itemNames = CommandLine.values("yc");
       		if ( itemNames != null && itemNames.size() > 0 )
       			parseItems(itemNames, DomainTypes.YIELD_CURVE, false, false, itemList);
    		
       		itemNames = CommandLine.values("sl");
       		if ( itemNames != null && itemNames.size() > 0 )
       			parseItems(itemNames, DomainTypes.SYMBOL_LIST, false, false, itemList);
    		
       		itemNames = CommandLine.values("sld");
       		if ( itemNames != null && itemNames.size() > 0 )
       			parseItems(itemNames, DomainTypes.SYMBOL_LIST, false, true, itemList);   

            if (itemList.size() == 0 && !CommandLine.hasArg("tunnel"))
            {
            	ItemArg itemArg = new ItemArg(DomainTypes.MARKET_PRICE, defaultItemName, false);
            	itemList.add(itemArg);            	
            }
            
            if ( tsServiceName() == null || tsServiceName().equals(""))
                connectionArg.tsService(connectionArg.service());
       		
            connectionArg.itemList(itemList);
       		
    		String value = CommandLine.value("publisherInfo");
    		if (value!= null) 
    		{
    			String [] pieces = value.split(",");
        		
    			if( pieces.length > 1 )
    			{
    				publisherId = pieces[0];
    				
    				try
    				{
    					Integer.parseInt(publisherId);
    				}
    				catch(Exception e)
    				{
    					System.err.println("Error loading command line arguments:\t");
    					System.out.println("publisherId within publisherinfo should be an integer number");
    					System.out.println("Consumer exits...");
    					System.exit(CodecReturnCodes.FAILURE);    					
    				}
    				publisherAddress = pieces[1];            	
    			}
    		}            
        }
        catch (IllegalArgumentException ile)
        {
            System.err.println("Error loading command line arguments:\t");
            System.err.println(ile.getMessage());
            System.err.println();
            System.err.println(CommandLine.optionHelpString());
            System.out.println("Consumer exits...");
            System.exit(CodecReturnCodes.FAILURE);
        }
		
        return true;
	}
        
	String serviceName()
	{
		return CommandLine.value("s");
	}
//APIQA	
	private int parseSocketConnectionArgs(String[] args, int argOffset)
	{
		int retCode = -1;
		ConnectionArg connectionArg = null;
		
		if ((args.length-1) >= argOffset+3)
		{
			if (args[argOffset+1].contains(":") && !args[argOffset+2].contains(":"))
			{
				String[] tokens = args[argOffset+1].split(":");
				if (tokens.length == 3)
				{
					connectionArg = new ConnectionArg();
					connectionArg.connectionType(ConnectionTypes.SOCKET);
					if (!args[argOffset+2].startsWith("-"))
					{
    				    connectionArg.service(args[argOffset+2]);
    				    retCode = argOffset + 3;
					}
					else
					{
					    retCode = argOffset + 2;
					}
					connectionArg.hostname(tokens[0]);
					connectionArg.port(tokens[1]);
					connectionArg.service(tokens[2]);
					connectionList.add(connectionArg);
					System.out.println("\nParsing connection arguments as server configuration...");
					System.out.println("HostName " +  connectionArg.hostname());
					System.out.println("Port " +  connectionArg.port());
					System.out.println("ServiceName " +  connectionArg.service() + "\n");
				}
				else
				{
					return retCode;
				}
			}
			else
			{
				return retCode;
			}
		}

		return retCode;
	}
		
	String tsServiceName()
	{
		return CommandLine.value("tsServiceName");
	}	

	List<ConnectionArg> connectionList()
	{
		return connectionList;
	}

	String userName()
	{
		return CommandLine.value("u");			
	}

	boolean enableView()
	{
		return CommandLine.booleanValue("view");
	}

	boolean enablePost()
	{
		return CommandLine.booleanValue("post");
	}

	boolean enableOffpost()
	{
		return CommandLine.booleanValue("offpost");
	}

    String publisherId()
    {
    	return publisherId;
    }
    
    String publisherAddress()
    {
    	return publisherAddress;
    }
    
	boolean enableSnapshot()
	{
		return CommandLine.booleanValue("snapshot");
	}

	int runtime()
	{
		return CommandLine.intValue("runTime");
	}
	
	boolean enableXmlTracing()
	{
		return CommandLine.booleanValue("x");
	}

	boolean enableEncrypted()
	{
		return enableEncrypted;
	}
	
	boolean enableHttp()
	{
		return enableHttp;
	}

	boolean enableProxy()
	{
		return CommandLine.booleanValue("proxy");
	}	
	
	String proxyHostname()
	{
		return CommandLine.value("ph");
	}

	String proxyPort()
	{
		return CommandLine.value("pp");
	}	
	
	String proxyUsername()
	{
		return CommandLine.value("plogin");
	}

	String proxyPassword()
	{
		return CommandLine.value("ppasswd");
	}	
		
	String proxyDomain()
	{
		return CommandLine.value("pdomain");
	}

	String krbFile()
	{
		return CommandLine.value("krbfile");
	}	
	
	String keyStoreFile()
	{
		return CommandLine.value("keyfile");
	}

	String keystorePassword()
	{
		return CommandLine.value("keypasswd");
	}	
		
    boolean cacheOption()
	{
		return cacheOption;
	}
	
	int cacheInterval()
	{
		return cacheInterval;
	}
	
	String authenticationToken()
    {
		return CommandLine.value("at");
    }

	String authenticationExtended()
    {
		return CommandLine.value("ax");
    }

	String applicationId()
    {
		return CommandLine.value("aid");
    }
	
	int itemCount()
	{
		return itemList.size();
	}
	
	public List<ItemInfo> itemList()
	{
		return itemList;
	}
		
	private void parseItems(List<String> itemNames, int domain, boolean isPrivateStream, boolean isSymbollistData, List<ItemArg> itemList)
	{
		for ( String itemName : itemNames)
		{    
			ItemArg itemArg = new ItemArg();
			itemArg.domain(domain);

            if (isPrivateStream) 
            	itemArg.enablePrivateStream(true);
            else 
            	itemArg.enablePrivateStream(false);
			if (isSymbollistData)
				itemArg.symbolListData(true);
			
			itemArg.itemName(itemName);
			itemList.add(itemArg);

		}
	}
//APIQA	
	void initialAllConnectionArgs()
	{
		for (int index = 1; index < connectionList.size(); index++)
		{
			connectionList.get(index).connectionType(connectionList.get(0).connectionType());
			connectionList.get(index).tunnel(connectionList.get(0).tunnel());
			connectionList.get(index).tunnelAuth(connectionList.get(0).tunnelAuth());
			connectionList.get(index).tunnelDomain(connectionList.get(0).tunnelDomain());
		}
	}
	
	List<ItemArg>  parseItemsFromFile()
	{
		//parsing item names from file to connectionArg
		if (itemFile.length() != 0)
		{
			XmlItemInfoList _xmlItemInfoList;
			_xmlItemInfoList = new XmlItemInfoList(itemCount);
			if (_xmlItemInfoList.parseFile(itemFile) == -1)
	        {
	        	System.out.printf("Failed to load item list from file '%s'.\n", itemFile);
				return null;
			}
			
			List<ItemArg> itemList = new ArrayList<ItemArg>();
			ItemArg itemArg;
			int numOfItem = _xmlItemInfoList.itemInfoList().length;
			for (int i=0; i<numOfItem; i++)
			{
				itemArg = new ItemArg();
				itemArg.domain(_xmlItemInfoList.itemInfoList()[i].domainType());
				itemArg.itemName(_xmlItemInfoList.itemInfoList()[i].name());
				itemList.add(itemArg);
			}
			
			return itemList;
		}
		return null;
	}
	
    public void addCommandLineArgs()
    {
        CommandLine.programName("WatchlistConsumer");
        CommandLine.addOption("mp", "For each occurrence, requests item using Market Price domain.");
        CommandLine.addOption("mbo", "For each occurrence, requests item using Market By Order domain. Default is no market by order requests.");
        CommandLine.addOption("mbp", "For each occurrence, requests item using Market By Price domain. Default is no market by price requests.");
        CommandLine.addOption("yc", "For each occurrence, requests item using Yield Curve domain. Default is no yield curve requests.");
        CommandLine.addOption("view", "Specifies each request using a basic dynamic view. Default is false.");
        CommandLine.addOption("post", "Specifies that the application should attempt to send post messages on the first requested Market Price item (i.e., on-stream). Default is false.");
        CommandLine.addOption("offpost", "Specifies that the application should attempt to send post messages on the login stream (i.e., off-stream).");
        CommandLine.addOption("publisherInfo", "Specifies that the application should add user provided publisher Id and publisher ipaddress when posting");       
        CommandLine.addOption("snapshot", "Specifies each request using non-streaming. Default is false(i.e. streaming requests.)");
        CommandLine.addOption("sl", "Requests symbol list using Symbol List domain. (symbol list name optional). Default is no symbol list requests.");
        CommandLine.addOption("sld", "Requests item on the Symbol List domain and data streams for items on that list.  Default is no symbol list requests.");
        CommandLine.addOption("h", defaultSrvrHostname, "Server host name");
        CommandLine.addOption("p", defaultSrvrPortNo, "Server port number");
        CommandLine.addOption("if", (String)null, "Interface name");
        CommandLine.addOption("s", defaultServiceName, "Service name");
        CommandLine.addOption("u", "Login user name. Default is system user name.");
        CommandLine.addOption("c", "Socket", "Specifies the connection type that the connection should use. Possible values are: 'Socket', 'http', 'encrypted'");

        CommandLine.addOption("runTime", defaultRuntime, "Program runtime in seconds");
        CommandLine.addOption("x", "Provides XML tracing of messages.");
        
        CommandLine.addOption("proxy", "Specifies that the application will make a tunneling connection (http or encrypted) through a proxy server, default is false");         
        CommandLine.addOption("ph", "", "Proxy server host name");
        CommandLine.addOption("pp", "", "Proxy port number");         
        CommandLine.addOption("plogin", "", "User Name on proxy server");
        CommandLine.addOption("ppasswd", "", "Password on proxy server");
        CommandLine.addOption("pdomain", "", "Proxy server domain");
        CommandLine.addOption("krbfile", "", "KRB File location and name");
        CommandLine.addOption("keyfile", "", "Keystore file location and name");
        CommandLine.addOption("keypasswd", "", "Keystore password");        
        
        CommandLine.addOption("tunnel", "", "(optional) enables consumer to open tunnel stream and send basic text messages");
        CommandLine.addOption("tsServiceName", "", "(optional) specifies the service name for tunnel stream messages (if not specified, the service name specified in -c/-tcp is used");
        CommandLine.addOption("tsAuth", "", "(optional) causes consumer to request authentication when opening a tunnel stream. This applies to basic tunnel streams");
        CommandLine.addOption("tsDomain", "", "(optional) specifes the domain a consumer uses when opening a tunnel stream. This applies to basic tunnel streams");
        
        CommandLine.addOption("at", "", "Specifies the Authentication Token. If this is present, the login user name type will be Login.UserIdTypes.AUTHN_TOKEN.");
        CommandLine.addOption("ax", "", "Specifies the Authentication Extended information.");
        CommandLine.addOption("aid", "", "Specifies the Application ID.");
        //APIAQ
		CommandLine.addOption("server", "Specified srvrHostName:srvrPortNo serviceName");
        CommandLine.addOption("itemFile", "Specified a file path for reading item requests");
        CommandLine.addOption("itemCount", "Specified how many items the app should read from the item request file");
        }	
}

File: CSharp/Eta/Applications/Examples/PHWatchlistConsumer/WatchlistConsumerConfig.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2025 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using LSEG.Eta.Codec;
using LSEG.Eta.Example.Common;
using LSEG.Eta.Rdm;
using LSEG.Eta.ValueAdd.Reactor;
using Microsoft.Extensions.Primitives;
using Newtonsoft.Json;
using Newtonsoft.Json.Serialization;

namespace LSEG.Eta.ValueAdd.WatchlistConsumer;

[ProgName("PHWatchlistConsumer")]
public partial class WatchlistConsumerConfig : CommandLine
{
    private const string DEFAULT_SERVICE_NAME = "DIRECT_FEED";

    private const string DEFAULT_ITEM_NAME = "TRI.N";

    private const string DEFAULT_CONFIG_FILE = "Config.json";

    private const int DEFAULT_RUNTIME = 600;

    private const int MAX_ITEMS = 128;

    private const int ITEMS_MIN_STREAM_ID = 5;

    private readonly List<ItemInfo> m_ItemList = new();

    private readonly ReactorConnectOptions m_ReactorConnectOpts = new();

    private readonly List<string> m_Errors = new();

    public IReadOnlyList<string> Errors => m_Errors;

    /// <summary>
    /// Watchlist consumer config item info.
    /// </summary>
    public class ItemInfo
    {
        public State State { get; set; } = new();

        public bool IsBatchStream { get; set; }

        public bool IsPrivateStream { get; set; }

        public bool SymbolListData { get; set; }

        public int StreamId { get; set; }

        public string? Name { get; set; }

        public string? ServiceName { get; set; }

        public DomainType Domain { get; set; }
    }

    public void AddItem(string itemName, string serviceName, DomainType domainType)
    {
        if (ItemList.Count >= MAX_ITEMS)
        {
            m_Errors.Add($"Config Error: Example only supports up to {MAX_ITEMS} items ");
            return;
        }
        ItemInfo itemInfo = new()
        {
            Domain = domainType,
            Name = itemName,
            ServiceName = serviceName,
            StreamId = ITEMS_MIN_STREAM_ID + ItemList.Count
        };
        m_ItemList.Add(itemInfo);
    }

    public ItemInfo? GetItemInfo(int streamId)
    {
        if (streamId > 0)
        {
            if (streamId >= ITEMS_MIN_STREAM_ID && streamId < ItemList.Count + ITEMS_MIN_STREAM_ID)
                return ItemList[streamId - ITEMS_MIN_STREAM_ID];
            else
                return null;
        }
        else
        {
            return null;
        }
    }

    public WatchlistConsumerConfig() => Init();

    private void Init()
    {
        m_ReactorConnectOpts.Clear();
        try
        {
            AddCommandLineArgs();
            ParseArgs();
            if (MarketPriceItemNames?.Count > 0)
                ParseItems(MarketPriceItemNames, DomainType.MARKET_PRICE);

            if (MarketByOrderItemNames?.Count > 0)
                ParseItems(MarketByOrderItemNames, DomainType.MARKET_BY_ORDER);

            if (MarketByPriceItemNames?.Count > 0)
                ParseItems(MarketByPriceItemNames, DomainType.MARKET_BY_PRICE);

            if (m_ItemList.Count == 0)
            {
                AddItem(DEFAULT_ITEM_NAME, ServiceName!, DomainType.MARKET_PRICE);
            }
        }
        catch (Exception exception)
        {
            m_Errors.Add("Error loading command line arguments:\t");
            m_Errors.Add(exception.Message);
        }

        try
        {
            if (ConfigJson is not null)
            {
                string json = File.ReadAllText(ConfigJson!);
                var jsonSerializerSettings = new JsonSerializerSettings
                {
                    ContractResolver = new CamelCasePropertyNamesContractResolver()
                };
                Config = JsonConvert.DeserializeObject<Json>(json, jsonSerializerSettings);
            }
        }
        catch (Exception exception)
        {
            m_Errors.Add($"Error loading {nameof(ConfigJson)}:");
            m_Errors.Add(exception.Message);
        }
    }

    [Option("s", DEFAULT_SERVICE_NAME, "Default service name for requests. This will be used for dictionary requests as well as any items that do not have a default service specified in -mp, -mbo, or -mbp.")]
    public string? ServiceName { get; private set; }

    [Option("u", "Login user name. Default is system user name.", Aliases = new string[] { "uname" })]
    public string? UserName { get; private set; }

    [Option("passwd", "Password for the user name.")]
    public string? Password { get; private set; }

    [Option("runTime", DEFAULT_RUNTIME, "Program runtime in seconds", Aliases = new string[] { "runtime" })]
    public int Runtime { get; private set; }

    [Option("x", "Provides XML tracing of messages.")]
    public bool EnableXmlTracing { get; private set; }

    [Option("clientId", "Specifies the client Id for Refinitiv login V2, or specifies a unique ID with login V1 for applications making the request to EDP token service, this is also known as AppKey generated using an AppGenerator.")]
    public string? ClientId { get; private set; }

    [Option("clientSecret", "Specifies the associated client Secret with a provided clientId for V2 logins.")]
    public string? ClientSecret { get; private set; }

    [Option("tokenURLV2", "Specifies the token URL for V2 token oauthclientcreds grant type.")]
    public string? TokenUrlV2 { get; private set; }

    [Option("tokenScope", "", "Specifies the token scope.")]
    public string? TokenScope { get; private set; }

    [Option("serviceDiscoveryURL", "Specifies the service discovery URL.")]
    public string? ServiceDiscoveryURL { get; private set; }

    [Option("mp", "For each occurrence, requests item using Market Price domain.")]
    public List<string>? MarketPriceItemNames { get; private set; }

    [Option("mbo", "For each occurrence, requests item using Market By Order domain. Default is no market by order requests.")]
    public List<string>? MarketByOrderItemNames { get; private set; }

    [Option("mbp", "For each occurrence, requests item using Market By Price domain. Default is no market by price requests.")]
    public List<string>? MarketByPriceItemNames { get; private set; }

    [Option("at", "", "Specifies the Authentication Token. If this is present, the login user name type will be Login.UserIdTypes.AUTHN_TOKEN.")]
    public string? AuthenticationToken { get; private set; }

    [Option("ax", "", "Specifies the Authentication Extended information.")]
    public string? AuthenticationExtended { get; private set; }

    [Option("aid", "", "Specifies the Application ID.")]
    public string? ApplicationId { get; private set; }

    [Option("audience", "", "Optionally specifies the audience used with V2 JWT logins")]
    public string? Audience { get; private set; }

    [Option("jwkFile", "Specifies the file location containing the JWK encoded private key for V2 logins.")]
    public string? JwkFile { get; private set; }

    [Option("configJson", DEFAULT_CONFIG_FILE, "Json config file path", Aliases = new string[] { "config" })]
    public string? ConfigJson { get; private set; }

    [Option("restProxyHost", "Proxy server host name for REST requests")]
    public string? RestProxyHost { get; private set; }

    [Option("restProxyPort", "Proxy port number for REST requests")]
    public string? RestProxyPort { get; private set; }

    [Option("restProxyUsername", "User Name on proxy server for REST requests")]
    public string? RestProxyUsername { get; private set; }

    [Option("restProxyPasswd", "Password on proxy server for REST requests")]
    public string? RestProxyPassword { get; private set; }

    [Option("reconnectAttemptLimit", -1, "Specifies the maximum number of reconnection attempts", Aliases = new[] { "reconnectCount", "reconnectLimit" })]
    public int ReconnectAttempLimit { get; private set; }

    [Option("reconnectMinDelay", 500, "Specifies minimum delay (in milliseconds) between reconnection attempts")]
    public int ReconnectMinDelay { get; private set; }

    [Option("reconnectMaxDelay", 6000, "Specifies maximum delay (in milliseconds) between reconnection attempts")]
    public int ReconnectMaxDelay { get; private set; }

    [Option("ioctlInterval", 0, "Specifies time interval (in second) before IOCtl function is invoked. O indicates that function won't be invoked. Default is 0")]
    public int IoctlInterval { get; private set; }

    [Option("fallBackInterval", 0, "Specifies time interval (in second) in application before Ad Hoc Fallback function is invoked. O indicates that function won't be invoked. Default is 0", Aliases = new[] { "fallbackInterval" })]
    public int FallBackInterval { get; private set; }

    [Option("ioctlEnablePH", "Enables Preferred host feature. Possible values are true/false")]
    public bool? IoctlEnablePH { get; private set; }

    [Option("ioctlConnectListIndex", "Specifies the preferred host as the index in the connection list")]
    public int? IoctlConnectListIndex { get; private set; }

    [Option("ioctlDetectionTimeInterval", "Specifies time interval (in second) to switch over to a preferred host. 0 indicates that the detection time interval is disabled")]
    public int? IoctlDetectionTimeInterval { get; private set; }

    [Option("ioctlDetectionTimeSchedule", "Specifies Cron time format to switch over to a preferred host")]
    public string? IoctlDetectionTimeSchedule { get; private set; }

    [Option("restEnableLog", false, "Enables REST logging message")]
    public bool EnableRestLogging { get; private set; }

    [Option("restLogFileName", "Specifies REST logging file stream name")]
    public string? RestLogFileName { get; private set; }

    internal Json? Config { get; private set; }

    public int ItemCount => ItemList.Count;

    public IReadOnlyList<ItemInfo> ItemList => m_ItemList;

    public bool IoctlOverridesConfigPreferredHost =>
        (IoctlEnablePH is not null) ||
        (IoctlConnectListIndex is not null && IoctlConnectListIndex != Config?.PreferredHost?.ConnectListIndex) ||
        (IoctlDetectionTimeInterval is not null && IoctlDetectionTimeInterval != Config?.PreferredHost?.DetectionTimeInterval) ||
        (IoctlDetectionTimeSchedule is not null && IoctlDetectionTimeSchedule != Config?.PreferredHost?.DetectionTimeSchedule);

    private void ParseItems(List<string> itemNames, DomainType domain)
    {
        foreach (string itemName in itemNames)
        {
            StringTokenizer tokenizer = new(itemName, new[] { ':' });
            string name;
            string serviceName;
            var segments = tokenizer.ToArray();
            if (segments.Length == 1 && segments[0].HasValue)
            {
                AddItem(segments[0].Value!, ServiceName!, domain);
            }
            else if (segments.Length == 2 && segments[0].HasValue && segments[1].HasValue)
            {
                serviceName = segments[0].Value!;
                name = segments[1].Value!;
                AddItem(name, serviceName, domain);
            }
            else
            {
                m_Errors.Add($"Invalid item name {itemName}.  This needs to be either <item_name> or <service_name>:<item_name");
            }
        }
    }
}
